Using math::SpeedLimiter on the diff_drive controller.

LolaSegura opened the pull request:
Signed-off-by: LolaSegura <lsegura@ekumenlabs.com>

## Summary
This replace the diff_drive controller speed limiter class for the one implemented in ignition math. Part of #810. 

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

francocipollone commented on src/systems/diff_drive/DiffDrive.cc:
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
   {
-    minVel = _sdf->Get<double>("min_velocity");
-    hasVelocityLimits = true;
+    double minVel = _sdf->Get<double>("min_velocity");
```
nit: it can be `const`

francocipollone commented on src/systems/diff_drive/DiffDrive.cc:
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
   {
-    maxVel = _sdf->Get<double>("max_velocity");
-    hasVelocityLimits = true;
+    double maxVel = _sdf->Get<double>("max_velocity");
```
nit: it can be `const`

francocipollone commented on src/systems/diff_drive/DiffDrive.cc:
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
   {
-    minAccel = _sdf->Get<double>("min_acceleration");
-    hasAccelerationLimits = true;
+    double minAccel = _sdf->Get<double>("min_acceleration");
```
nit: it can be `const`

francocipollone commented on src/systems/diff_drive/DiffDrive.cc:
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
   {
-    maxAccel = _sdf->Get<double>("max_acceleration");
-    hasAccelerationLimits = true;
+    double maxAccel = _sdf->Get<double>("max_acceleration");
```
nit: it can be `const`

francocipollone commented on src/systems/diff_drive/DiffDrive.cc:
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
   {
-    minJerk = _sdf->Get<double>("min_jerk");
-    hasJerkLimits = true;
+    double minJerk = _sdf->Get<double>("min_jerk");
```
nit: it can be `const`

francocipollone commented on src/systems/diff_drive/DiffDrive.cc:
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
   {
-    maxJerk = _sdf->Get<double>("max_jerk");
-    hasJerkLimits = true;
+    double maxJerk = _sdf->Get<double>("max_jerk");
```
nit: it can be `const`

LolaSegura replied:
Addressed all.

francocipollone reviewed (commented):
LGTM! I just left a couple of nits.

I will add @scpeters as a reviewer as well

scpeters commented:
this is looking good!

it's helpful to cross-reference issues from pull requests, so I would add that this is "Part of https://github.com/ignitionrobotics/ign-gazebo/issues/810" in the pull request description

scpeters commented:
I noticed that codecov is complaining about test coverage. We are not strict about requiring code coverage, but I think we can easily increase this for this test. I believe the `diff_drive` system has the following integration test:

* https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo3/test/integration/diff_drive_system.cc

The test uses several different world files (search for `test/worlds` to find [lines like this](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo3_3.8.0/test/integration/diff_drive_system.cc#L203)) such as [test/worlds/diff_drive.sdf](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo3_3.8.0/test/worlds/diff_drive.sdf). This world has a plugin that [specifies max_acceleration and max_velocity](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo3_3.8.0/test/worlds/diff_drive.sdf#L230-L231), but doesn't specify `min_acceleration` or `min_velocity`. If we added `min_acceleration` of `-1` and `min_velocity` of `0.5`, that will improve the coverage and add symmetry to the test. To get even more coverage and without changing the dynamics of the system we could add a `min_jerk` value of `-Inf` and a `max_jerk` value of `Inf` (I think that will work).

Since these are easy changes to improve code coverage, do you mind giving them a try? Thanks!

LolaSegura commented:
> If we added `min_acceleration` of -1 and `min_velocity` of 0.5, that will improve the coverage and add symmetry to the test.

@scpeters  here when you mentioned the `min_velocity` did you meant to write -0.5? Because otherwise it would take the same value as the `max_velocity`.

scpeters commented:
> > If we added `min_acceleration` of -1 and `min_velocity` of 0.5, that will improve the coverage and add symmetry to the test.
> 
> @scpeters here when you mentioned the `min_velocity` did you meant to write -0.5? Because otherwise it would take the same value as the `max_velocity`.

thanks for catching that; I meant to write `-0.5`

francocipollone reviewed (changes requested):
ptal at the comments. Also, run `codecheck` to match style.

francocipollone commented:
@LolaSegura . I've tried it locally.
 - It seems that just adding the `min_jerk` and `max_jerk` properties are causing a miss behavior in the diff drive when moving forward(probably backward too).
 - Leaving those properties commented out the test moving forward works fine.
 - It is failing only one test when moving backward:
 ```
 Expected: (expectedLowerPosition.Y()) < (expectedGreaterPosition.Y()), actual: 0.710173 vs 1.40412e-12
 ```
 Re-check the desire angle you are publishing.

LolaSegura commented:
>If we added min_acceleration of -1 and min_velocity of 0.5 -0.5, that will improve the coverage and add symmetry to the test

I added the `<min_velocity>` and `<min_acceleration>` to the `diff_drive.sdf` and the `diff_drive_custom_topics.sdf`. With this the test its working fine. 

> To get even more coverage and without changing the dynamics of the system we could add a min_jerk value of -Inf and a max_jerk value of Inf (I think that will work).

I try adding this limits as well, but I encounter a problem because when being parsed the limits where interpreted as zero, causing the test to fail. An option would be setting the `<max_jerk>` with a big value and the `<min_jerk>` with a small one to get more test coverage.

francocipollone commented:
It is ready for review @scpeters

scpeters reviewed (commented):
> > To get even more coverage and without changing the dynamics of the system we could add a min_jerk value of -Inf and a max_jerk value of Inf (I think that will work).
> 
> I try adding this limits as well, but I encounter a problem because when being parsed the limits where interpreted as zero, causing the test to fail. An option would be setting the `<max_jerk>` with a big value and the `<min_jerk>` with a small one to get more test coverage.

I'll look into this briefly

LolaSegura commented on test/integration/diff_drive_system.cc:
```diff
@@ -54,145 +54,170 @@ class DiffDriveTest : public ::testing::TestWithParam<int>
+      else
+      {
+        EXPECT_GT(v, -0.5);
+        EXPECT_GT(a, -1);
```
Addressed both comments.

scpeters commented:
> > To get even more coverage and without changing the dynamics of the system we could add a min_jerk value of -Inf and a max_jerk value of Inf (I think that will work).
> 
> I try adding this limits as well, but I encounter a problem because when being parsed the limits where interpreted as zero, causing the test to fail. An option would be setting the `<max_jerk>` with a big value and the `<min_jerk>` with a small one to get more test coverage.

thanks for testing this; I just looked into it a bit deeper and found that it is a parsing issue with `libsdformat` (https://github.com/osrf/sdformat/issues/261). We can leave out the jerk tests for now
